##########################################################################################
# Let a haplotype be a sequence of alleles observed at polymorphic site within a window
# e.g. For a window with 1 triallelic and 2 biallelic SNP, 6 possible haplotypes: 00,01,10,12,20,21
# Let m denote the number of windows, indexed by t = 1 to m
# Let Kt be the number of haplotypes with non-zero frequency per window
# Let max(Kt) be the maximum over t = 1,...,m
# Due to a separate objective, examples below inc. only windows with diversity >= 0.5 
#
# Other columns: 
# S.hap.div = window wise measure of haplotype diversity based on Senegal
# FG.hap.div = window wise measure of haplotype diversity based on French Guiana
##########################################################################################
rm(list = ls())

# Here is an example of how Emily presented frequencies of haplotypes in a data frame
# I think, frequencies were generated by calculate_haplotype_frequencies_per_window 
# Then filtered, excluding windows whose diversoty < 0.5
load('../RData/FromEmily/top.windows.hap.frequencies.FG.RData') 
head(top.windows.hap.frequencies.FG) # 

# I then reformat the fequencies above s.t. they can be stored in sparse m by max(Kt) matrix: 
load(file = '../RData/freq_FG_S.RData')
head(frequencies_FG) # for each row, entries 1:Kt != 0, Kt+1:max(Kt) = 0
head(rowSums(frequencies_FG)) # all rows sum to one

# In addition to frequencies, to simulate data I use 
# chrom: extracted from chromosome
# pos: centre point of window
# dt: distance between centre point of current and preceding window
# Keff_FG, Keff_S, Med_Keff, Med_div, Med_div_01 are all simple measures of diversity, 
# transformed in diffeent ways
load('../RData/data_FG_S.RData')
head(data_)  


